Yosemite/Yosemite/Stores/ShipmentStore.swift

```diff
@@ -0,0 +1,90 @@
+import Foundation
+import Networking
+import Storage
+import CocoaLumberjack
+
+
+// MARK: - ShipmentStore
+//
+public class ShipmentStore: Store {
+
+    /// Shared private StorageType for use during then entire ShipmentStore sync process
+    ///
+    private lazy var sharedDerivedStorage: StorageType = {
+        return storageManager.newDerivedStorage()
+    }()
+
+    /// Registers for supported Actions.
+    ///
+    override public func registerSupportedActions(in dispatcher: Dispatcher) {
+        dispatcher.register(processor: self, for: ShipmentAction.self)
+    }
+
+    /// Receives and executes Actions.
+    ///
+    override public func onAction(_ action: Action) {
+        guard let action = action as? ShipmentAction else {
+            assertionFailure("ShipmentStore received an unsupported action")
+            return
+        }
+
+        switch action {
+        case .synchronizeShipmentTrackingData(let siteID, let orderID, let onCompletion):
+            synchronizeShipmentTrackingData(siteID: siteID, orderID: orderID, onCompletion: onCompletion)
+        }
+    }
+}
+
+
+// MARK: - Services!
+//
+private extension ShipmentStore {
+
+    func synchronizeShipmentTrackingData(siteID: Int, orderID: Int, onCompletion: @escaping (Error?) -> Void) {
+        let remote = ShipmentsRemote(network: network)
+        remote.loadShipmentTrackings(for: siteID, orderID: orderID) { [weak self] (shipmentTrackingData, error) in
+            guard let readOnlyShipmentTrackingData = shipmentTrackingData else {
+                onCompletion(error)
+                return
+            }
+
+            self?.upsertShipmentTrackingDataInBackground(siteID: siteID, orderID: orderID, readOnlyShipmentTrackingData: readOnlyShipmentTrackingData) {
+                onCompletion(nil)
+            }
+        }
+    }
+}
+
+
+// MARK: - Persistence
+//
+extension ShipmentStore {
+
+    /// Updates (OR Inserts) the specified ReadOnly ShipmentTracking Entities into the Storage Layer *in a background thread*. onCompletion will be called
+    /// on the main thread!
+    ///
+    func upsertShipmentTrackingDataInBackground(siteID: Int, orderID: Int, readOnlyShipmentTrackingData: [Networking.ShipmentTracking], onCompletion: @escaping () -> Void) {
+        let derivedStorage = sharedDerivedStorage
+        derivedStorage.perform {
+            for readOnlyTracking in readOnlyShipmentTrackingData {
```

Comment:
Just curious, but what is the reason to do this in a for loop instead of map? For what it is worth, I think it is better to do it the way it's done here, as I tend to avoid doing any work that has side effects within a map closure, but I was wondering if that is also a project-wide convention.

Reply:
Good question! I guess this is a project-level convention, but I am not opposed to using a map here either FWIW. 😄